src/digests.rs

```diff
@@ -0,0 +1,211 @@
+//! # Multihash
+//!
+//! Implementation of [multihash](https://github.com/multiformats/multihash) in Rust.
+//!
+//! A `Multihash` is a structure that contains a hashing algorithm, plus some hashed data.
+//! A `MultihashRef` is the same as a `Multihash`, except that it doesn't own its data.
+//!
+
+use std::convert::TryFrom;
+
+use bytes::{BufMut, Bytes, BytesMut};
+use num_traits::cast::FromPrimitive;
+use unsigned_varint::{decode as varint_decode, encode as varint_encode};
+
+use crate::errors::{DecodeError, DecodeOwnedError};
+use crate::hashes::Code;
+
+/// Represents a valid multihash.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Multihash {
+    bytes: Bytes,
+}
+
+impl Multihash {
+    /// Verifies whether `bytes` contains a valid multihash, and if so returns a `Multihash`.
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Multihash, DecodeOwnedError> {
+        if let Err(err) = MultihashRef::from_slice(&bytes) {
+            return Err(DecodeOwnedError {
+                error: err,
+                data: bytes,
+            });
+        }
+        Ok(Multihash {
+            bytes: Bytes::from(bytes),
+        })
+    }
+
+    /// Returns the bytes representation of the multihash.
```

Comment:
I apologize if I'm misunderstanding this but as far as I understand this would return bytes without consuming `Multihash` as `to_vec` creates a copy (isn't it). This seems to contradict expectation from [`into_bytes` in `String`s](https://doc.rust-lang.org/std/string/struct.String.html#method.into_bytes).

If my assumption is correct this method probably should either be removed or changed so it consumes multihash to match String behavior.

Reply:
I think this is correct. A method named `into_` carries with it the expectation that it consumes self and can thus perform the conversion in a cheap way.

For naming guidelines, see https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv